### ISSUE TITLE
📊 Analysis: Eagle3 Architecture & Implementation

### ISSUE THREAD

**Issue by KilJaeeun** [KilJaeeun/eagle-hallu-shift-idea#3]
# 📊 Eagle3 Architecture Analysis

Eagle3 논문 및 구현체 분석 결과를 정리합니다.

---

## 1. Multi-Layer Feature Fusion

Eagle3는 target model의 **3개 shallow layers (0, 1, 2)**를 사용합니다:

```python
# EAGLE/eagle/traineagle3/cnets.py (line 717-720)
hidden_states0 = outs.hidden_states[0]  # Embedding layer
hidden_states1 = outs.hidden_states[1]  # Layer 1
hidden_states2 = outs.hidden_states[2]  # Layer 2
hidden_states = torch.cat((hidden_states0, hidden_states1, hidden_states2), dim=-1)
```

### 왜 Shallow Layers인가?
1. **빠른 계산**: 깊은 레이어까지 가지 않아도 됨
2. **Syntactic/local information**: 초기 레이어가 next-token prediction에 충분
3. **Memory efficiency**: 중간 레이어 저장 불필요

---

## 2. Feature Combination Method

**Concatenation + FC Layer**:
```python
# Concatenation → [batch, seq, hidden_size * 3]
hidden_states = torch.cat((hidden_states0, hidden_states1, hidden_states2), dim=-1)

# FC for dimensionality reduction
self.fc = nn.Linear(self.hidden_size * 3, self.hidden_size, bias=False)
hidden_states = self.fc(hidden_states)
```

**특징**:
- Simple concatenation (no weighted sum)
- Single linear projection (no bias)
- Output: original hidden_size

---

## 3. Training Objective

**Knowledge Distillation Loss (KL-divergence based)**:
```python
# Lines 841-855
target_p = nn.Softmax(dim=2)(target_head)      # Target probabilities
target_p = target_p.detach()                    # No gradient

logits = self.lm_head(hidden_states_out)        # Draft logits
out_logp = nn.LogSoftmax(dim=2)(logits)         # Log softmax

plogp = target_p * out_logp
loss = -torch.sum(position_mask * plogp, 2).mean()
```

**Multi-step Training**:
```python
# Exponential weighting for 7 positions
ploss_weight = [0.8 ** i for i in range(len(plosses))]
ploss = sum([ploss_weight[i] * plosses[i] for i in range(len(plosses))])
```

---

## 4. Draft Model Architecture

**Single Decoder Layer**:
```python
# cnets.py line 491
self.midlayer = LlamaDecoderLayeremb(config)  # Only 1 layer!
```

**Implication**: Layer-to-layer dynamics 분석이 draft 내부에서는 불가능

---

## 5. Acceptance Rate Calculation

```python
# Lines 857-859
acces.append(
    ((logits.argmax(-1) == target_p.argmax(-1)) * position_mask.squeeze(-1)).sum().item() 
    / (loss_mask.sum().item() + 1e-6)
)
```

**Metric**: argmax match rate (draft argmax == target argmax)

---

## 6. Known Limitations (Our Opportunity)

| Limitation | Description | Our Solution |
|------------|-------------|--------------|
| Static layer fusion | 단순 concat, dynamics 무시 | Layer delta prediction |
| No uncertainty signal | 불확실한 구간 인식 못함 | Attention entropy conditioning |
| Error accumulation | Later positions에서 성능 저하 | **Main target** |
| Single draft layer | Draft 내부 dynamics 분석 불가 | Target dynamics로 supervision |

---

## 7. Key Files

| File | Purpose |
|------|---------|
| `EAGLE/eagle/traineagle3/cnets.py` | Training model definition |
| `EAGLE/eagle/traineagle3/main.py` | Training loop |
| `EAGLE/eagle/model/cnets.py` | Inference model |
| `EAGLE/eagle/model/utils.py` | Acceptance evaluation |

---

## References
- [EAGLE-3 Paper (arXiv 2503.01840)](https://arxiv.org/abs/2503.01840)
- [EAGLE GitHub](https://github.com/SafeAILab/EAGLE)
